README.md
Phox Signal Toggle / Brake Sensor combo
------------
The Phox Signal Toggle and Brake Sensor combo (what a terrible name) can be paired with a Phox Tail Light to display left and right turn signal animations and brake animation.

Development
------------
There are a few tools required to develop this stuff:
* \*nix box with make, gcc, the usual stuff
* [makeEspArduino makefile](https://github.com/plerup/makeEspArduino)
* [ArduinoEsp](https://github.com/esp8266/Arduino)

First, clone this repo. Be sure to include `--recursive` so that it will pull down submodules:

    git clone --recursive https://github.com/Phoxlights/PhoxSignalBrakeCombo.git

Then update the makefile to point to your local copy of the makeEspArduino and the ArduinoEsp repo:

    ESP_MAKE=$(HOME)/src/makeEspArduino
    ESP_LIBS=$(HOME)/src/ArduinoEsp/libraries

Finally, hook up your esp8266 and try `make upload`. Will it work? Probably!

Releasing
-----------
Use the [git flow](http://nvie.com/posts/a-successful-git-branching-model/) release process. 

Release checklist:
* Bump to BIN_VERSION in whatever sketch (TODO - automatically pull this from VERSION)
* Build a new bin and put it in `build`. This can be done with `make build`.
* Update `CHANGELOG`
* Make sure submodules are pinned to a release tag
* *After* release, bump the version in `VERSION` and push to develop


CHANGELOG
0.0.1
===========
### Changes
* made this thing!

### Config
button pin 14
status pin 2
toggle left pin 4
toggle right pin 0
brake pin 5
connects to network with ssid phoxlightpriv, pass phoxlightpriv
hostname phoxsigbrake
sync mode (button hold), connects to ssid phoxlight, pass phoxlight
ota hostname phoxsigbrakeota
event version 2
event port 6767
db ver 3

### Commands
Run Mode:
* ping
* who
Sync mode:
* setdefaultconfig
* setnetworkmode
* registerconfirm


src/signalbrakeconfig.cpp
#include <Esp.h>
#include <objstore.h>
#include <identity.h>
#include "signalbrakeconfig.h"

#define DEFAULT_CONFIG_ID 1

// default config values
SignalBrakeConfig defaultConfig = {
    PRIVATE_SSID,
    PRIVATE_PASS,
    HOSTNAME,
    CONNECT,
};

static Identity id = {
    .model = 1000,
    .serial = ESP.getChipId(),
    .bin = BIN_VERSION,
    .eventVer = EVENT_VER,
    .dbVer = DB_VER
};

// only one config for now
static SignalBrakeConfig config;

SignalBrakeConfig * getConfig(){
    return &config;
}

Identity * getIdentity(){
    return &id;
}

PrivateNetworkCreds getPrivateCreds(){
    PrivateNetworkCreds privateCreds;
    strcpy(privateCreds.ssid, config.ssid);
    strcpy(privateCreds.pass, config.pass);
    return privateCreds;
}

int loadConfig(){
    // load from fs
    objStoreInit(DB_VER);
    if(!objStoreGet("signalbrake", DEFAULT_CONFIG_ID, &config, sizeof(SignalBrakeConfig))){
        // store defaults
        Serial.println("signalbrake config not found, storing defaults");
        if(!writeDefaultConfig()){
            return 0;
        }
        // reload config now that we have defaults
        loadConfig();
    }
    return 1;
}

int writeConfig(SignalBrakeConfig * c){
    if(!objStoreUpdate("signalbrake", DEFAULT_CONFIG_ID, c, sizeof(SignalBrakeConfig))){
        Serial.println("failed to write config");
        return 0;
    }
    return 1;
}

static void logComponent(Identity * id){
    Serial.printf("\
    {\n\
        model: %i,\n\
        serial: %i,\n\
        bin: %i,\n\
        eventVer: %i,\n\
        dbVer: %i,\n\
    }\n",
    id->model, id->serial, id->bin, id->eventVer, id->dbVer);
}

void logConfig(SignalBrakeConfig * c){
    Serial.printf("\
config: {\n\
    ssid: %s,\n\
    hostname: %s,\n\
    networkMode: %s,\n\
    }\n",
    c->ssid, c->hostname,
    c->networkMode == 0 ? "CONNECT" : c->networkMode == 1 ? "CREATE" : "OFF");
}

int writeDefaultConfig(){
    // delete existing signalbrake config
    Serial.println("wiping signalbrake config");
    objStoreWipe("signalbrake");

    int id = objStoreCreate("signalbrake", &defaultConfig, sizeof(SignalBrakeConfig));
    if(!id){
        Serial.println("failed to write default config");
        return 0;
    }
    Serial.printf("wrote default signalbrake config, with id %i\n", id);
    return 1;
}

/*
// TODO - linked list isntead of array
int registerComponent(Identity * component){
    Identity * cur;

    Serial.printf("attempting to register component\n");
    logComponent(component);

    int i;
    for(i = 0; i < MAX_COMPONENTS; i++){
        cur = &(config.components[i]);

        // this one exists already, so update
        if(cur->serial == component->serial){
            Serial.printf("updating existing component with serial %i\n", cur->serial);
            identityCopy(cur, component);
            return 1;
        }

        if(!identityIsDefined(cur)){
            // hey, heres a free slot!
            break;
        }
    }

    // dont register more than MAX_COMPONENTS
    if(i >= MAX_COMPONENTS - 1){
        Serial.printf("cannot register more components, already maxed out\n");
        return 0;
    }

    // add the component
    Serial.printf("registering component to slot %i\n", i);
    cur = &(config.components[i]);
    identityCopy(cur, component);

    return 1;
}
*/


src/signalbrakeconfig.h
#ifndef SIGNALBRAKECONFIG_H
#define SIGNALBRAKECONFIG_H

#include <network.h>
#include <identity.h>

#define DB_VER 3
#define EVENT_VER 2
#define EVENT_PORT 6767
#define BRAKE_PIN 5
#define TOGGLE_LEFT_PIN 4
// TODO - dont use GPIO 0
#define TOGGLE_RIGHT_PIN 0
#define BUTTON_PIN 14
#define STATUS_PIN 2
// TODO - get BIN_VERSION from VERSION file
#define BIN_VERSION 2

// 192.168.4.1 17082560
// 192.168.43.20 338405568
#define SERVER_IP_UINT32 17082560

#define HOSTNAME "phoxsigbrake"
#define OTA_HOSTNAME "phoxsigbrakeota"

#define PUBLIC_SSID "phoxlight"
#define PUBLIC_PASS "phoxlight"

// TODO - obtain these from controller
#define PRIVATE_SSID "phoxlightpriv"
#define PRIVATE_PASS "phoxlightpriv"

typedef struct SignalBrakeConfig {
    char ssid[SSID_MAX];
    char pass[PASS_MAX];
    char hostname[HOSTNAME_MAX];
    NetworkMode networkMode;
} SignalBrakeConfig;

typedef struct PrivateNetworkCreds {
    char ssid[SSID_MAX];
    char pass[PASS_MAX];
} PrivateNetworkCreds;

SignalBrakeConfig * getConfig();
Identity * getIdentity();
PrivateNetworkCreds getPrivateCreds();

int loadConfig();
int writeConfig(SignalBrakeConfig * c);
void logConfig(SignalBrakeConfig * c);
int writeDefaultConfig();

//int registerComponent(Identity * component);
// TODO - removeComponent

#endif
